### ISSUE TITLE
non-ASCII request body characters may get corrupted in a non UTF-8 environment

### ISSUE THREAD

**Issue by Russell-** [jayway/rest-assured#677]
def String serialize(ObjectMapperSerializationContext context) {
        def object = context.getObjectToSerialize()
        JsonEncoding jsonEncoding = getEncoding(context.getCharset())
        def mapper = createJackson2ObjectMapper(object.getClass(), context.getCharset())
        def stream = new ByteArrayOutputStream()
        JsonGenerator jsonGenerator = mapper.getJsonFactory().createJsonGenerator(stream, jsonEncoding)
        mapper.writeValue(jsonGenerator, object)
        return stream.toString()
    }

Shown above is the serialize method from com.jayway.restassured.internal.mapping.Jackson2Mapper.  I'm experienceing an issue with it.  The problem arises if I have a non-ASCII character in the body.  I'm running with:

     RestAssured.config = new RestAssuredConfig().encoderConfig(new EncoderConfig("UTF-8","UTF-8"));

If I run a JUnit test in an enviroment where Charset.defaultCharset() returns UTF-8 all works correctly.  The same test when run in a Windows environment where Charset.defaultCharset() returns windows-1252 results in the issue.

Let's walk through the code with an attribute in the request body having a value of "Ñ" without the quotes (corresponding to the Unicode U+00D1 code point).  The "charset" in the context object (ObjectMapperSerializationContextImpl) on line 51 is UTF-8.  As such, "jsonEncoding" is UTF-8 based and after line 55 (mapper.writeValue()) is executed the stream byte array contains C3 91 for "Ñ". The two bytes C3 91 are the correct UTF-8 represenatation for "Ñ".  The issue arises in the next line - stream.toString().  Absent any charset specified to toString() (there is a variation of toString() which takes a charset name argument) the code to construct a String from a byte array will use Charset.defaultCharset() which in this case is windows-1252.  So it interprets C3 91 as being in windows-1252 encoding and not UTF-8 (which it is) and results in serialize() returning the following 2 Unicode characters: 00C3 and 2018 (conforming to https://en.wikipedia.org/wiki/Windows-1252).  These two characters (00C3 and 2018) are seen in the Controller method for that request body attribute.

In short "stream.toString()" needs to take the charset corresponding to the encoding used by mapper.writeValue - in my case since mapper.writeValue() writes the bytes out in UTF-8 stream.toString() should read it as UTF-8 and not use the default charset which may or may not be UTF-8.

**Comment by johanhaleby** [comment]
Thanks for a detailed error report. I've tried taking charset into account when generating the string. I've now deployed a new snapshot that hopefully fixes this. Please try version `2.9.1-SNAPSHOT` after having added the following Maven repository:

```xml
<repositories>
        <repository>
            <id>sonatype</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
            <snapshots />
        </repository>
</repositories>
```

**Comment by Russell-** [comment]
Thanks, Johan, for taking care of this so promptly.  Your change addresses the problem. We're not allowed to consume SNAPSHOTs at our site.  When do you expect the next release to be?

**Comment by johanhaleby** [comment]
Thanks for trying it out. Unfortunately I cannot tell when it's going to be released. I'd like to look into a few other issues as well and since I'm (mostly) working on RA on my spare time I cannot give any promises. I'll close this issue for now.
